src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,167 @@
+                head.prev = newNode;
+                head = newNode;
+                size++;
+                return;
+            }
+            Node<T> currentNode = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (indexIsValid(index)) {
+            return getNodeByIndex(index).item;
+        } else {
+            throw new IndexOutOfBoundsException("Index is not valid");
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (indexIsValid(index)) {
+            T temp = getNodeByIndex(index).item;
+            getNodeByIndex(index).item = value;
+            return temp;
+        } else {
+            throw new IndexOutOfBoundsException("Index is not valid");
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index is not valid");
+        } else {
+            if (size == 1 && index == 0) {
+                size--;
+                Node<T> removedNode = tail;
+                tail = head = null;
+                return removedNode.item;
+            }
+            if (index == size - 1) {
+                size--;
+                Node<T> removedNode = tail;
+                tail = tail.prev;
+                tail.next.prev = tail.next = null;
+                return removedNode.item;
+            }
+            if (index == 0) {
+                size--;
+                Node<T> removedNode = head;
+                head = head.next;
+                head.prev.next = head.prev = null;
+                return removedNode.item;
+            }
+            Node<T> currentNode = getNodeByIndex(index);
+            currentNode.prev.next = currentNode.next;
+            currentNode.next.prev = currentNode.prev;
+            size--;
+            return currentNode.item;
+        }
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        if (getIndexByValue(object) == -1) {
+            return false;
+        } else {
+            remove(getIndexByValue(object));
+            return true;
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
```

Comment:
```suggestion
    private Node<T> getNodeByIndex(int index) {
        Node<T> currentNode = head;
        for (int i = 0; i < index; i++) {
            currentNode = currentNode.next;
        }
        return currentNode;
    }
```
you mas use non only move from head, also you can use move from tail. It help wen List wery big